bonsai-core/src/main/scala/com/stripe/bonsai/IndexedBitSet.scala

```diff
@@ -0,0 +1,306 @@
+package com.stripe.bonsai
+
+import scala.annotation.switch
+import scala.collection.immutable.BitSet
+import scala.collection.mutable.{ Builder, ArrayBuilder }
+
+/**
+ * IndexedBitSet is an immutable set for storing non-negative integer
+ * values.
+ * 
+ * This data structure stores non-negative integers. It has O(1)
+ * lookup and rank operations. It also supports O(polylog n) select
+ * operations, which are not currently implemented.
+ *
+ * To support fast rank, this bitset requires ~1.37n bits, rather than
+ * the n bits that other bitsets (e.g. `scala.BitSet`) would use.
+ *
+ * Unlike some immutable structures, there is no form of structural
+ * sharing here. Thus, the only way to "modify" one of these
+ * structures is to build a new one that has the desired modifications
+ * (which will not be efficient if done frequently).
+ */
+final class IndexedBitSet(
+  val bits: Array[Int],
+  val length: Int,
+  val level2Start: Int,
+  val rawBitsStart: Int
+) {
+  import IndexedBitSet.rankWord
+
+  /**
+   * Test if `i` is contained in this bitset.
+   *
+   * The value `i` is expected to be non-negative.
+   */
+  def apply(i: Int): Boolean = {
```

Comment:
So, with that said... To go in another direction, if we were just encoding the 32-bit indices / bits in the long - that is, just using the longs to encode 2 logical words - then I think we'd probably make things a bit more complex, but I don't think we'd see any perf gains, since the Ints would be laid out side-by-side in memory anyways. Does that make sense?